app/oauth.go

```diff
@@ -765,17 +765,27 @@ func (a *App) GetAuthorizationCode(w http.ResponseWriter, r *http.Request, servi
 
 	redirectURI := siteURL + "/signup/" + service + "/complete"
 
-	authURL := endpoint + "?response_type=code&client_id=" + clientId + "&redirect_uri=" + url.QueryEscape(redirectURI) + "&state=" + url.QueryEscape(state)
+	authURL, sErr := url.Parse(endpoint)
+	if sErr != nil {
+		return "", model.NewAppError("GetAuthorizationCode.ParseEndpoint", "api.user.get_authorization_code.endpoint_parse.app_error", nil, "failed to parse authentication endpoint URL", http.StatusBadRequest).Wrap(sErr)
+	}
+	queryParams := authURL.Query()
+	queryParams.Set("response_type", "code")
+	queryParams.Set("client_id", clientId)
+	queryParams.Set("redirect_uri", url.QueryEscape(redirectURI))
+	queryParams.Set("state", url.QueryEscape(state))
```

Comment:
You don't need to escape when using URL builder because it'll [eventually encode each param](https://github.com/golang/go/blob/master/src/net/url/url.go#L980). If you do it beforehand, you're double encoding, which may lead to an incorrect URL.

Reply:
@plant99 Maybe we could update the tests to catch this?

Reply:
Thanks for raising this @kostaspt , I've left these two fields which were encoded with `%20` as they were. For others, I removed the second encoding.